strawberryfields/utils/post_processing.py

```diff
@@ -0,0 +1,271 @@
+# Copyright 2019-2020 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+This module provides functions for post-processing samples.
+"""
+import numpy as np
+
+
+def _product_for_modes(samples, modes=None):
+    """Getting the product of samples across modes.
+Args:
+        samples (ndarray): the photon number samples with a shape
+            of (shots, modes)
+        modes (Sequence): a flat sequence containing indices of modes to get
+            the expectation value for
+
+    Returns:
+        ndarray: product of the samples across modes
+    """
+    modes = np.array(modes)
+    selected_samples = samples[:, modes]
+    return np.prod(selected_samples, axis=1)
+
+
+def _check_samples(samples):
+    """Validation check for the input samples.
+
+    Checks include data types checks and dimension of the input samples.
+
+    Args:
+        samples (ndarray): the photon number samples with a shape of (shots,
+            modes)
+    """
+    if not isinstance(samples, np.ndarray) or samples.ndim != 2:
+        raise Exception("Samples needs to be represented as a two dimensional NumPy array.")
+
+
+def _check_modes(samples, modes):
+    """Validation checks for the input modes.
+
+    Checks include data types checks, checks for the dimension of the inputs
+    and the validity of the modes specified.
+
+    Args:
+        samples (ndarray): the photon number samples with a shape
+            of (shots, modes)
+        modes (Sequence): the input modes to get the expectation value for, a
+            flattened sequence
+    """
+    num_modes = samples.shape[1]
+    flattened_sequence_indices_msg = (
+        "The input modes need to be specified as a flattened sequence of non-negative integers!"
+    )
+
+    modes = np.array(modes)
+
+    # Checking if modes is a valid flattened sequence of indices
+    try:
+        non_index_modes = modes[(np.mod(modes, 1) != 0) | (modes < 0)]
+    except Exception:
+        raise Exception(flattened_sequence_indices_msg)
```

Comment:
Interesting approach here 🤔 

What exception is likely to be raised here? Often, it is better to catch a specific exception, rather than all.

Reply:
Wondering if it might be better to have separate if statements, as used below?

Reply:
Thanks, valid point! The reason why this was in a `try-except` block is because it is supposed to handle cases when the array does not contain numerical values but rather other objects (e.g. lists). Added a comment on this and adjusted the `try-except` to except for `TypeError` (this is raised in such cases) and to raise this type of error.

Further changed to raising `ValueError` for other cases.